src/chol_derivative.h

```diff
@@ -0,0 +1,181 @@
+    for (int j = 0; j < theta.size(); j++) {
+      matrix<Type> ld2 = chol_d2_vec.segment( (j * theta.size() + i) * n_visits_sq, n_visits_sq).matrix();
+      matrix<Type> ld1_j = chol_d1_vec.segment(j * n_visits_sq, n_visits_sq).matrix();
+      ld2.resize(n_visits, n_visits);
+      ld1_j.resize(n_visits, n_visits);
+      auto ld2_lt = ld2 * l.transpose();
+      auto ld1_ld1j = ld1 * ld1_j.transpose();
+      auto sigma_d2_ij = ld2_lt + ld2_lt.transpose() + ld1_ld1j + ld1_ld1j.transpose();
+      ret_d2.block((i * theta.size() + j) * n_visits, 0, n_visits, n_visits) = sigma_d2_ij;
+    }
+  }
+  ret["derivative1"] = ret_d1;
+  ret["derivative2"] = ret_d2;
+  return ret;
+}
+// struct chols is created to get the derivatives with cache.
+// The main reason to have it is that we nearly always have duplicated visits
+// and the inverse of a matrix is calculation expensive. In addition, we can save
+// the resource needed for select matrix calculations.
+template <class Type>
+struct chols {
+  std::map<std::vector<int>, matrix<Type>> inverse_cache;
+  std::map<std::vector<int>, matrix<Type>> sigmad1_cache;
+  std::map<std::vector<int>, matrix<Type>> sigmad2_cache;
+  std::map<std::vector<int>, matrix<Type>> sigma_cache;
+  std::map<std::vector<int>, matrix<Type>> sigma_inverse_d1_cache;
+  std::map<std::vector<int>, Eigen::SparseMatrix<Type>> sel_mat_cache;
+  std::string cov_type;
+  std::vector<int> full_visit;
+  int n_visits;
+  int n_theta; // theta.size()
+  vector<Type> theta;
+  matrix<Type> chol_full;
+  chols() {
+    // this default constructor is needed because the use of `[]`.
+  }
+  chols(vector<Type> theta, int n_visits, std::string cov_type): n_visits(n_visits), cov_type(cov_type), full_visit(std::vector<int>(n_visits)) {
+    this->theta = theta;
+    for (int i = 0; i < n_visits; i++) {
+      this->full_visit[i] = i;
+    }
+    this->n_theta = theta.size();
+    std::map<std::string, tmbutils::matrix<Type>> allret = derivatives<Type>(this->n_visits, this->cov_type, this->theta);
+    matrix<Type> sigma_d1 = allret["derivative1"];
+    matrix<Type> sigma_d2 = allret["derivative2"];
+    matrix<Type> l = allret["chol"];
+    this->chol_full = l;
+    this->sigmad1_cache[this->full_visit] = sigma_d1;
+    this->sigmad2_cache[this->full_visit] = sigma_d2;
+    this->sigma_cache[this->full_visit] = allret["chol"] * allret["chol"].transpose();
+    this->inverse_cache[this->full_visit] = (this->sigma_cache[this->full_visit]).inverse();
+    this->sel_mat_cache[this->full_visit] = get_select_matrix<Type>(full_visit, this->n_visits);
+  }
+  Eigen::SparseMatrix<Type> get_sel_mat(std::vector<int> visits) {
+    if (this->sel_mat_cache.count(visits) > 0) {
+      return this->sel_mat_cache[visits];
+    } else {
+      this->sel_mat_cache[visits] = get_select_matrix<Type>(visits, this->n_visits);
+      return this->sel_mat_cache[visits];
+    }
+  }
+  matrix<Type> get_sigma_derivative1(std::vector<int> visits) {
```

Comment:
please also here on top of the internal functions/methods please add comment lines to explain.